[issue-1077] Display logic refactoring

yromanyshyn opened the pull request:
resolves #1077 
resolves #627 

As was said in the issue description currently, the code responsible for the check&suite displays is not clean and very hard to maintain. From my point of view in order to clean it up we need to do next things:
1. Decouple display logic from `CheckResult` and `SuiteResult`
2. Decouple different display approaches from each other (html, widget) 

I suggest the next set of the abstractions in order to achieve this 

```python

# Serializer - class that implements serialization logic for the value of type `T` into some particular format 

class Serializer(t.Protocol[T]):
    value: T
    def __init__(self, value: T, **kwargs):
        self.value = value

class HtmlSerializer(Serializer[T]):
    def serialize(self, **kwargs) -> str:
        ...

class JsonSerializer(Serializer[T]):
    def serialize(self, **kwargs) -> t.Dict[t.Any, t.Any]:
        ...

class WidgetSerializer(Serializer[T]):
    def serialize(self, **kwargs) -> Widget:
        ...

# ============

# Presentation - a facade class for a set of serializers, it should not contain 
# implementation of a serialization logic, its main task is to instantiate instances 
# of serializers and to provide an interface to them. 

class Presentation(t.Protocol[T]):
    value: T

    def __init__(self, value: T, **kwargs):
        self.value = value

    def to_json(self, **kwargs) -> t.Dict[t.Any, t.Any]:
        raise NotImplementedError()

    def to_html(self, **kwargs) -> str:
        raise NotImplementedError()

    def to_widget(self, **kwargs) -> Widget:
        raise NotImplementedError()

# ============

# It is also totally logical to add a `Deserializer` type that will contain 
# logic opposite to the logic  `Serializer` but currently I am not sure about this idea
```

For more details take a look at the code.

Note: currently this PR is in the draft state but it already contains basic logic for `CheckResult` and `SuiteResult`

matanper commented on deepchecks/core/presentation/check_result/html.py:
```diff
@@ -0,0 +1,166 @@
+    ) -> t.List[str]:
+        output = [cls.header()]
+
+        for item in display:
```
why do all the handles in here transform to HTML? If later I want to convert it to json?

yromanyshyn replied:
the reason why I added this class is to make it possible to reuse HTML serialization functionality in widget serializer without using inheritance which makes code flow harder to follow and understand.

If it is useful with JSON serializer I will adapt it and will make it more flexible.

yromanyshyn replied:
I created a separate protocol called  `ABCDisplayItemsHandler` that is now used everywhere in the subpackage when we need to deal with display items

matanper reviewed (commented):
Very nice! I like that.

1. I'm not sure why we need the `Presentation` facade objects. I feel they don't add anything as we still need to manually choose "to_html/json/widget".
2. I feel like there is still some middleware missing that will handle the "structural" stuff. Some common logic under the "serialize" function seems like they can be in a common place for all serializers type (of a given class)

yromanyshyn commented:
@matanper 

> I feel like there is still some middleware missing that will handle the "structural" stuff. Some common logic under the "serialize" function seems like they can be in a common place for all serializers type (of a given class)

probably, it will become more clear when I implement JSON and wandb serialization

>I'm not sure why we need the Presentation facade objects. I feel they don't add anything as we still need to manually choose "to_html/json/widget".

1. Its main role is to aggregate a set of serializers into one 'point'
2. IMO it can be useful with their help to use a 'strategy pattern', example below (not sure if it is important or needed currently/)

```python

# CheckResult (or SuiteResult) implementation

class CheckResult:
    DEFAULT_PRESENTATION = PresentationType

    def __init__(self, ..., presentation: Type[Presentation[CheckResult]] = None):
         ...
         presentation_type = presentation or self.DEFAULT_PRESENTATION
         self._presentation = presentation_type(self)

    def to_json(self):
         return self._presentation.to_json()
    
    def __ipython_display__(self):
         dipslay(self._presentation.to_widget())
   
    ...

# Checks and Suite impl

class SomeCheck(SingleDatasetBaseCheck):
     def __init__(self, ..., presentation: Type[Presentation[CheckResult]] = None):
         self.presentation_type = presentation
     def run(self):
         return CheckResult(..., presentation=self.presentation_type)

class SomeSuite(BaseSuite):
     def __init__(self, ..., presentation: Type[Presentation[SuiteResult]] = None):
         ...
         self.presentation_type = presentation
     def run(self):
         ...
         return SuiteResult(..., presentation=self.presentation_type)

```

matanper commented:
@y
romanyshyn my point is you still have to call the correct format manually.
```
class CheckResult:
    DEFAULT_PRESENTATION = PresentationType

    def __init__(self, ..., presentation: Type[Presentation[CheckResult]] = None):
         ...
         presentation_type = presentation or self.DEFAULT_PRESENTATION
         self._presentation = presentation_type(self)

    def to_json(self):
         return self._presentation.to_json() OR CheckJsonSerializer().format(self)
    
    def __ipython_display__(self):
         dipslay(self._presentation.to_widget())  OR CheckWIdgetSerializer().format(self)
```
So if the call is not so different, it feels like just adding redundant code

matanper commented:
@yromanyshyn my point is you still have to call the correct format manually.
```
class CheckResult:
    DEFAULT_PRESENTATION = PresentationType

    def __init__(self, ..., presentation: Type[Presentation[CheckResult]] = None):
         ...
         presentation_type = presentation or self.DEFAULT_PRESENTATION
         self._presentation = presentation_type(self)

    def to_json(self):
         return self._presentation.to_json() OR CheckJsonSerializer().format(self)
    
    def __ipython_display__(self):
         dipslay(self._presentation.to_widget())  OR CheckWIdgetSerializer().format(self)
```
So if the call is not so different, it feels like just adding redundant code

yromanyshyn commented:
@matanper 

>@yromanyshyn my point is you still have to call the correct format manually.

yes I understand that, but my point is a little bit different - it is about flexibility and giving users the possibility to redefine that behavior.

Anyway, after some consideration, I also think that it is redundant at the current moment. Therefore I removed that trait and renamed the subpackage.

yromanyshyn commented:
attaching screenshots to demonstrate how display looks after refactoring

Note: not all bugs are fixed yet

[zip with an example notebook](https://github.com/deepchecks/deepchecks/files/8520200/nb-examples.zip)


Suite Result to widget
------------------------------

1.
![image](https://user-images.githubusercontent.com/71635444/164194450-0d32e361-fd8a-4d7e-a693-31445f1c5da3.png)

2.
![image](https://user-images.githubusercontent.com/71635444/164194596-500071ad-6c8a-456b-84ee-a77589d01704.png)

3.
![image](https://user-images.githubusercontent.com/71635444/164194694-226ee4a1-8bb4-4e57-b030-b92e69093a1e.png)

Suite Result to wandb
-----------------------------

1.
![image](https://user-images.githubusercontent.com/71635444/164195092-1d7c33b5-563a-4c45-8845-bcc3782b5547.png)

2.
![image](https://user-images.githubusercontent.com/71635444/164195785-ca3a18f2-5472-42d7-904b-b408cd4887e9.png)

3.
![image](https://user-images.githubusercontent.com/71635444/164195873-7915da57-bf32-4086-9131-b1fb51ed0ed3.png)

matanper commented:
@yromanyshyn I don't think this is needed. The callable is not there to return arbitrary object, it is a hack specifically for matplotlib, since we want to allow write checks with matplotlib but it got problematic architecture (plots are global and are shown when they are created, it's not possible to return figure object and display it when you like)

yromanyshyn commented:
@matanper 

> @yromanyshyn I don't think this is needed. 

hm/ do not know, this will allow lazy figure/display creation and that could speed up checks a little bit in some cases

> problematic architecture (plots are global and are shown when they are created, it's not possible to return figure object and display it when you like)

it is possible to have a figure object instance, matplotlib provides two coding styles: 'OOP style' and 'Global API style' ([docs](https://matplotlib.org/stable/tutorials/introductory/usage.html#coding-styles))

yromanyshyn commented:
@matanper 

> > problematic architecture (plots are global and are shown when they are created, it's not possible to return figure object and display it when you like)
> 
> it is possible to have a figure object instance, matplotlib provides two coding styles: 'OOP style' and 'Global API style' ([docs](https://matplotlib.org/stable/tutorials/introductory/usage.html#coding-styles))

oh, I see what you mean

yromanyshyn commented:
@matanper 

> plots are global and are shown when they are created,

there is actually a way to prevent this behavior. Matplot provides a [plt.ioff](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.ioff.html) function for this.

Anyway, I implemented callable items handling in the same way how it was done previously. (only with help of `plt.ioff`)

matanper commented:
> @matanper
> 
> > plots are global and are shown when they are created,
> 
> there is actually a way to prevent this behavior. Matplot provides a [plt.ioff](https://matplotlib.org/stable/api/_as_gen/matplotlib.pyplot.ioff.html) function for this.
> 
> Anyway, I implemented callable items handling in the same way how it was done previously. (only with help of `plt.ioff`)

We don't want to change the behavior of matplotlib for the user, so if he got interactive mode on, we don't want to turn it off

yromanyshyn commented:
@matanper 

>We don't want to change the behavior of matplotlib for the user, so if he got interactive mode on, we don't want to turn it off and change it environment

I am using it as a context manager, interactive mode is disabled only during a display function call

```python
with plt.ioff():
    item()
    return common.read_matplot_figures()
```

noamzbr commented on spelling-allowlist.txt:
```diff
@@ -74,3 +74,10 @@ jsonpickle
+requirejs
+CDN
+CustomNotebookRenderer
+pretify
```
Correct spelling is prettify and pretty, so no need for the allowlist

yromanyshyn commented:
UPDATE. What is left to do:

- [ ] Write tests for `SuiteResult` serialization.
- [ ] Remove (replace) old display logic.

noamzbr commented on spelling-allowlist.txt:
```diff
@@ -74,4 +74,10 @@ jsonpickle
 dtype
 dataloader
 KDE
+renderes
```
I think it's renders (without an e) but not sure what was the meaning

yromanyshyn replied:
it was complaining about this line 
```python
class CustomNotebookRenderer(plotly_renderes.NotebookRenderer):
```
`plotly_renderes` is a plotly submodule 

actually, this class can be deleted but I would like to keep it for now
(anyway it is commented)

yromanyshyn replied:
oh) it is actually an alias, will fix)

yromanyshyn commented:
@ItayGabbay 
what do you think if we add the next config options for the isort step? (example below)
[isort documentation](https://pycqa.github.io/isort/docs/configuration/multi_line_output_modes.html)

also, maybe let's start using [black](https://github.com/psf/black) to autoformat code?

```
[tool.isort]
multi_line_output = 3
lines_after_imports = 2
line_length = 120

# multi_line_output example
#
# 3 - Vertical Hanging Indent
#
# from third_party import (
#    lib1,
#    lib2,
#    lib3,
#    lib4,
# )
```

matanper commented on deepchecks/core/serialization/check_result/html.py:
```diff
@@ -0,0 +1,325 @@
+            <html>
+            <head><meta charset="utf-8"/></head>
+            <body>
+                {''.join([requirejs, plotlyjs, *sections])}
```
we should put requirejs and plotlyjs in the head for standard?

yromanyshyn replied:
what it will change? 
also putting a script tag into the head is not a standard practice as from my point/

matanper commented on deepchecks/core/serialization/check_result/json.py:
```diff
@@ -0,0 +1,217 @@
+        conditions_table = ''
+
+    if display:
+        additional_output = ['<h5><b>Additional Outputs</b></h5>']
```
Html in json serializer?

matanper commented on deepchecks/core/serialization/check_result/json.py:
```diff
@@ -0,0 +1,217 @@
+        with warnings.catch_warnings():
+            warnings.simplefilter(action='ignore', category=FutureWarning)
+            conditions_table = DataFrameSerializer(df.style.hide_index()).serialize()
+            conditions_table = f'<h5><b>Conditions Summary</b></h5>{conditions_table}'
```
Html in json serializer?

matanper commented:
Overall it looks good. @yromanyshyn It will be great if you can write a quick guide on "how to add new display type" so it will be easier for us to understand how much works that requires. For example I want to add new Markdown display, how much logic I have to think about in terms of the structuring of stuff, etc, and how much is already built-in from the architecture

yromanyshyn commented:
@matanper 

> Overall it looks good. @yromanyshyn It will be great if you can write a quick guide on "how to add new display type" so it will be easier for us to understand how much works that requires. For example I want to add new Markdown display, how much logic I have to think about in terms of the structuring of stuff, etc, and how much is already built-in from the architecture

In general nothing special, but be sure that you comply with next points:
1. If you need a new type of a serializer (like for markdown for example) you need to add a dedicated trait for it in `...serialization.abc.py` module
2. there is only one requirement for the serializer type - to have a `serialize(self, **kwargs)` method
3. follow the structure of the `serialization/...` subpackage
4. when you need to deal with `CheckResult.display` items - use `...serialization.abc.ABCDisplayItemsHandler` for this. It will make sure that we deal with it, in the same manner, everywhere
5. if you need to reuse some logic from another serializer type, do not use inheritance, use composition instead (1. it of course depends on the context but is recommended 2. take a look at widget serializers to understand how it could be done)

In the case of a markdown serializer, it should be possible to reuse some logic from the HTML serializer, but in the current state, the HTML serializer will require some restructuring to separate HTML tags from plain text.

skeleton for the markdown serializer will look something like this:

```python

# === in ...serialization.abc.py

@runtime_checkable
class MardownSerializer(Serializer[T], Protocol):
    """To markdown serializer protocol."""
    def serialize(self, **kwargs) -> str:
        """Serialize into mardown."""
        ...

# === in ...serialization.check_result.mardown.py

class CheckResultSerializer(MardownSerializer[CheckResult]):
    def __init__(self, value: CheckResult, **kwargs):
        self.value = value
    def serialize(self, **kwargs) -> str: 
        ...
    def prepare_additional_output(self) - str:
        return DisplayItemsHandler.handle_display(self.value.display,)

class DisplayItemsHandler(ABCDisplayItemsHandler):
    
    @classmethod
    def handle_string(cls, item, index, **kwargs) -> str:
        ...
    @classmethod
    def handle_dataframe(cls, item, index, **kwargs) -> str:
        ...
    @classmethod
    def handle_callable(cls, item, index, **kwargs) -> str:
        figures = super().handle_callable(item, index, **kwargs)  # this will return list of open matplot figures
        ...
    @classmethod
    def handle_figure(cls, item, index, **kwargs) -> str:
        ...

# === in ../core.check_result.py

class CheckResult:
    def to_mardown(self) -> str:
        return CheckResultMarkdownSerializer(self).serialize()      

```

JKL98ISR commented on deepchecks/core/serialization/common.py:
```diff
@@ -0,0 +1,369 @@
+]
+
+
+# class CustomNotebookRenderer(plotly_renderers.NotebookRenderer):
```
please remove comment

JKL98ISR replied:
@yromanyshyn

yromanyshyn replied:
yup, removed

JKL98ISR commented on deepchecks/core/serialization/check_failure/wandb.py:
```diff
@@ -0,0 +1,65 @@
+
+try:
+    import wandb
+    from wandb.sdk.data_types.base_types.wb_value import WBValue
```
what is this?

yromanyshyn replied:
wandb is not a part of our default requirements, the user needs to install it separately if he wants to work with wandb serialization logic and we need to inform him about this as quickly as possible

JKL98ISR replied:
what is the WBValue?

yromanyshyn replied:
base type for all wandb data types - [docs](https://docs.wandb.ai/ref/python/data-types)

JKL98ISR commented on deepchecks/core/serialization/common.py:
```diff
@@ -0,0 +1,369 @@
+        return f'<script>{js}</script>'
+
+
+def plotlyjs_script(connected: bool = True) -> str:
```
why is it suddenly needed?

yromanyshyn replied:
plotly plots do not work without plotlyjs javascript library

JKL98ISR replied:
but we didn't do anything for it and it worked

yromanyshyn replied:
yes, because we have been using IPython.display and Plotly API for plots display that handles this under the hood
but in some cases, we need more precise control over the output and need to deal with this by ourselves

JKL98ISR replied:
The plotlyjs is according to the installed plotly version?

yromanyshyn replied:
yes

```python
...
from plotly.io._utils import plotly_cdn_url
from plotly.offline.offline import get_plotlyjs
...
plotly_cdn=plotly_cdn_url().rstrip(".js")
script=get_plotlyjs()
...
```

ItayGabbay reviewed (approved):
I love the decouple between the result and the display
